src/libsync/propagateuploadng.cpp

```diff
@@ -30,58 +30,207 @@
+    int _nrange = 0;
+    off_t *zbyterange = zsync_needed_byte_ranges(zs.get(), &_nrange, 0);
+    if (!zbyterange) {
+        abortWithError(SyncFileItem::NormalError, tr("Failed to get zsync byte ranges."));
+        return;
+    }
+
+    qCDebug(lcZsyncPut) << "Number of ranges:" << _nrange;
+
+    /* If we have no ranges then we have equal files and we are done */
+    if (_nrange == 0) {
+        propagator()->reportFileTotal(*_item, 0);
+        free(zbyterange);
+        finalize();
+        return;
+    }
+
+    /**
+     * If the `_item->size` is smaller than remote file then zbyterange is expected to have ranges that are
+     * outside it's size. This is because of the simplicity of the current upload algorithm in `zsync`. It
+     * currently will just return all the differences between the two files up to the remote file
+     * size. This is because of the case of `Moved` blocks within a file, blocks that are at the end of
+     * the remote file might be useful to the local file, they might just need to be moved them earlier in
+     * the file.
+     */
+    int totalBytes = 0;
+    for (int i=0; i<_nrange; i++) {
+        UploadRangeInfo rangeinfo = { (quint64)zbyterange[(2*i)], (quint64)zbyterange[(2*i)+1] - (quint64)zbyterange[(2*i)] + 1 };
+        if (rangeinfo.start < _item->_size) {
+            if (rangeinfo.start + rangeinfo.size > _item->_size)
+                rangeinfo.size = _item->_size - rangeinfo.start;
+            _rangesToUpload.append(rangeinfo);
+            totalBytes += rangeinfo.size;
+        }
+    }
+
+    /**
+     * _item->_size here is the local file size, where as zsync_file_length will provide the size
+     * of the remote item according to the zsync metadata downloaded. So if we have more bytes than
+     * remote then we must assume we have to upload them. This is related to the simple implementation
+     * for upload path today, but is an area for future work.
+     */
+    if (_item->_size > (quint64)zsync_file_length(zs.get())) {
+        UploadRangeInfo rangeinfo = { (quint64)zsync_file_length(zs.get()), (quint64)_item->_size - (quint64)zsync_file_length(zs.get()) };
+        _rangesToUpload.append(rangeinfo);
+        totalBytes += rangeinfo.size;
+    }
+    qCDebug(lcZsyncPut) << "Total bytes:" << totalBytes;
+    propagator()->reportFileTotal(*_item, totalBytes);
+
+    _bytesToUpload = totalBytes;
+    free(zbyterange);
+
+    doStartUploadNext();
+}
 
+void PropagateUploadFileNG::slotZsyncSeedFailed(const QString& errorString)
+{
+    qCCritical(lcZsyncPut) << errorString;
+
+    /* delete remote zsync file */
+    QUrl zsyncUrl = zsyncMetadataUrl(propagator(), _item->_file);
+    (new DeleteJob(propagator()->account(), zsyncUrl, this))->start();
+
+    abortWithError(SyncFileItem::NormalError, errorString);
+}
 
+/*
+State machine:
+
+  +---> doStartUpload()
+        isZsyncPropagationEnabled()?  +--+ yes +---> Download and seed zsync metadata and set-up new _rangesToUpload
```

Comment:
Should you not check the DB entry for resuming, first?
Maybe then the metadata or some additional state need to be kept.

Reply:
The ranges needed can't be known until the metadata is downloaded and seeded, the DB is checked after to see what's still needed/missing.